### PR TITLE
Cleaning up the test related classes 

### PR DESCRIPTION
**Pull request by rabbah**
This is a pass to clean up some of the type hierarchies that resulted from splitting of the tests between CLI and REST. 

This is not meant to be the end of the cleanup but a meaningful incremental checkpoint. 
In this PR:
- [x] removed redundancies between the rest and cli type hierarchies
- [x] removed unnecessary mixins
- [x] removed buried actor system in rest variant which results in 6 actor systems per instance
- [x] renamed the types to be clearer (imo)
- [x] removed obsolete tests
- [x] removed `return` uses
- [x] removed some uses of `var` that i came across
- [x] other scala-isms such as use of map.empty,  s/loop/map

Labeling as breaking out of caution since there was a lot of refactoring - I kept the top level `Wsk` type to minimize downstream disruptions.

## Related issue and scope

- [x] I opened an issue to propose and discuss this change (#3737)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [x] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [x] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [ ] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

File: tests/src/test/scala/common/RunCliCmd.scala [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import java.io.File
+
+import scala.collection.JavaConversions.mapAsJavaMap
+import scala.collection.mutable.Buffer
+import org.scalatest.Matchers
+import TestUtils._
+
+import scala.collection.mutable
+
+trait RunCliCmd extends Matchers {
```
Comment (rabbah) [inline tests/src/test/scala/common/RunCliCmd.scala]: this is not new code but the organization of the trait is different - namely the `cli` and `baseCommand` methods are grouped here and then used by `wskadmin` and `wsk` variants.

File: tests/src/test/scala/common/WskCliOperations.scala [hunk 2]
```diff
@@ -62,18 +65,17 @@ import whisk.utils.retry
...
  */
-class Wsk() extends RunWskCmd with BaseWsk {
-  override implicit val action = new WskAction
-  override implicit val trigger = new WskTrigger
-  override implicit val rule = new WskRule
-  override implicit val activation = new WskActivation
-  override implicit val pkg = new WskPackage
-  override implicit val namespace = new WskNamespace
-  override implicit val api = new WskApi
+class Wsk extends WskOperations {
```
Comment (rabbah) [inline tests/src/test/scala/common/WskCliOperations.scala]: I adopted the convention of `<Entity>Operations` for the base type, and prefixed `Cli` or `Rest` for the respective variants.

File: tests/src/test/scala/common/WskCliOperations.scala [hunk 3]
```diff
@@ -161,68 +162,11 @@ trait DeleteFromCollectionCLI extends BaseDeleteFromCollection {
   }
 }
 
-trait HasActivation {
```
Comment (rabbah) [inline tests/src/test/scala/common/WskCliOperations.scala]: This trait was moved.

File: tests/src/test/scala/common/rest/WskRestOperations.scala [hunk 4]
```diff
@@ -1186,34 +1170,34 @@ class WskRestApi extends RunWskRestCmd with BaseApi {
...
       parameters = parms,
       blocking = true,
       result = true,
       web = true,
       expectedExitCode = expectedExitCode)(wp)
-    return rr
   }
 }
 
-class RunWskRestCmd() extends FlatSpec with RunWskCmd with Matchers with ScalaFutures with WskActorSystem {
```
Comment (rabbah) [inline tests/src/test/scala/common/rest/WskRestOperations.scala]: that this is a class with an actor system that was extended by all the entity commands meant we allocated several instances of the actor system unnecessary. Further because this class isn't actually part of any test suite, the after all hook may not actually have been reached at all.

**Comment by rabbah** [comment]
note to reviewers: all the commits are incremental and can be reviewed independently.

**Comment by mdeuser** [comment]
@houshengbo - did you have any comments/concerns with these changes?

File: tests/src/test/scala/common/rest/WskRest.scala [hunk 5]
```diff
@@ -723,10 +720,7 @@ class WskRestActivation extends RunWskRestCmd with HasActivationRest with WaitFo
    * @return sequence of activations
    */
   def idsActivation(rr: RestResult): Seq[String] = {
-    val list = rr.getBodyListJsObject()
-    var result = Seq[String]()
-    list.foreach((obj: JsObject) => result = result :+ (RestResult.getField(obj, "activationId").toString))
-    result
+    rr.getBodyListJsObject.map(r => RestResult.getField(r, "activationId").toString)
```
Comment (mdeuser) [inline tests/src/test/scala/common/rest/WskRest.scala]: is a `.toSeq` needed to convert the `Vector[String]` to `Seq[String]` ?

File: tests/src/test/scala/common/rest/WskRest.scala [hunk 6]
```diff
@@ -1114,32 +1104,31 @@ class WskRestApi extends RunWskRestCmd with BaseApi {
                     expectedExitCode: Int = SUCCESS_EXIT,
                     cliCfgFile: Option[String] = None)(implicit wp: WskProps): RestResult = {
 
-    val parms = Map[String, JsValue]() ++ {
+    val parms = {
       basepathOrApiName map { b =>
         Map("basepath" -> b.toJson)
-      } getOrElse Map[String, JsValue]()
+      } getOrElse Map.empty
```
Comment (mdeuser) [inline tests/src/test/scala/common/rest/WskRest.scala]: is there a scala best practice preference (assuming a code block and method invocation implementations are equivalent)?

`basepathOrApiName.map(b => Map("basepath" -> b.toJson)).getOrElse(Map.empty)`

Comment (markusthoemmes) [reply tests/src/test/scala/common/rest/WskRest.scala]: There's none I believe, I tend to use your way of of writing @mdeuser because it feels more explicit, but that's just me. The compiler also tends to get funny with infix notation in some edge-cases, so I usually avoid that alltogether.

File: tests/src/test/scala/whisk/core/cli/test/ApiGwRestBasicTests.scala [hunk 7]
```diff
@@ -36,7 +36,7 @@ import common.WskProps
 @RunWith(classOf[JUnitRunner])
 abstract class ApiGwRestBasicTests extends BaseApiGwTests {
 
-  val clinamespace = wsk.namespace.whois()
+  lazy val clinamespace = wsk.namespace.whois()
```
Comment (mdeuser) [inline tests/src/test/scala/whisk/core/cli/test/ApiGwRestBasicTests.scala]: is it best practice to use `lazy` for abstract class property initialization?

Comment (rabbah) [reply tests/src/test/scala/whisk/core/cli/test/ApiGwRestBasicTests.scala]: required to avoid NPE when there are overrides.

File: tests/src/test/scala/whisk/core/cli/test/WskRestBasicUsageTests.scala [hunk 8]
```diff
@@ -48,14 +49,14 @@ import whisk.http.Messages
...
 
-  implicit val wskprops: common.WskProps = WskProps()
-  val wsk: common.rest.WskRest = new WskRest
+  implicit val wskprops = WskProps()
+  val wsk = new WskRestOperations
   val defaultAction: Some[String] = Some(TestUtils.getTestActionFilename("hello.js"))
   val usrAgentHeaderRegEx: String = """\bUser-Agent\b": \[\s+"OpenWhisk\-CLI/1.\d+.*"""
 
-  behavior of "Wsk CLI usage"
+  behavior of "Wsk API basic usage"
```
Comment (drcariel) [inline tests/src/test/scala/whisk/core/cli/test/WskRestBasicUsageTests.scala]: 👍 good catch - `val usrAgentHeaderRegEx: String = """\bUser-Agent\b": \[\s+"OpenWhisk\-CLI/1.\d+.*"""` is also unnecessary here - something else I missed to clean up when separating the tests tests.

**Comment by mdeuser** [comment]
@drcariel - are your downstream changes all set/ready for this to be merged?

**Comment by drcariel** [comment]
@mdeuser yes, I am ready for the dance 💃 🕺

**Comment by rabbah** [comment]
@mdeuser please use this commit message below when you squash and merge - I think for posterity I'd like to preserve all the commits in this PR so I am not going to squash myself.


```
Clean up wsk Rest and CLI interfaces.

- Remove unnecessary dependency injection and flatten type hiearchy some more.
- Remove burried ActorSystem which turned into several instances for every instantiated instance of the Rest 'wsk'.
- Remove swift tests.
- Rename test suite.
- Delete really dated sequence schema related tests.
- Remove return statements.
- Scala-isms and more code cleanup.
- Remove RunWskAdminCmd trait, move a trait and singleton around.
- Add deprecation warning to utility method used downstream (via log message).
- Allow wsk CLI path to be overriden.
- Allow for commands to on network error. Add test for run cmd with retry.
```

**Comment by mdeuser** [comment]
@rabbah correct.  maybe i misundersood
> I think for posterity I'd like to preserve all the commits in this PR

**Comment by rabbah** [comment]
I meant in the PR. Sorry for the confusion - that’s why I provided a commit message in the comment.
